dot/sync/worker_pool.go

```diff
@@ -16,217 +14,179 @@ import (
-
-	s.wg.Add(1)
-	go worker.run(workerQueue, &s.wg)
-
-	s.workers[who] = &syncWorker{
-		worker: worker,
-		queue:  workerQueue,
-	}
+	s.workers[who] = struct{}{}
 	logger.Tracef("potential worker added, total in the pool %d", len(s.workers))
+	return nil
 }
 
-// submitRequest given a request, the worker pool will get the peer given the peer.ID
-// parameter or if nil the very first available worker or
-// to perform the request, the response will be dispatch in the resultCh.
-func (s *syncWorkerPool) submitRequest(request *messages.BlockRequestMessage,
-	who *peer.ID, resultCh chan<- *syncTaskResult) {
-
-	task := &syncTask{
-		request:  request,
-		resultCh: resultCh,
+// submitRequests takes an set of requests and will submit to the pool through submitRequest
+// the response will be dispatch in the resultCh
+func (s *syncWorkerPool) submitRequests(tasks []*syncTask) []*syncTaskResult {
+	if len(tasks) == 0 {
+		return nil
 	}
 
-	// if the request is bounded to a specific peer then just
-	// request it and sent through its queue otherwise send
-	// the request in the general queue where all worker are
-	// listening on
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
 
-	if who != nil {
-		syncWorker, inMap := s.workers[*who]
-		if inMap {
-			if syncWorker == nil {
-				panic("sync worker should not be nil")
-			}
-			syncWorker.queue <- task
-			return
-		}
+	pids := maps.Keys(s.workers)
+	workerPool := make(chan peer.ID, len(pids))
+	for _, worker := range pids {
+		workerPool <- worker
 	}
 
-	// if the exact peer is not specified then
-	// randomly select a worker and assign the
-	// task to it, if the amount of workers is
-	var selectedWorkerIdx int
-	workers := maps.Values(s.workers)
-	nBig, err := rand.Int(rand.Reader, big.NewInt(int64(len(workers))))
-	if err != nil {
-		panic(fmt.Errorf("fail to get a random number: %w", err))
+	failedTasks := make(chan *syncTask, len(tasks))
+	results := make(chan *syncTaskResult, len(tasks))
+
+	var wg sync.WaitGroup
+	for _, task := range tasks {
+		wg.Add(1)
+		go func(t *syncTask) {
+			defer wg.Done()
+			executeTask(t, workerPool, failedTasks, results)
+		}(task)
 	}
-	selectedWorkerIdx = int(nBig.Int64())
-	selectedWorker := workers[selectedWorkerIdx]
-	selectedWorker.queue <- task
-}
 
-// submitRequests takes an set of requests and will submit to the pool through submitRequest
-// the response will be dispatch in the resultCh
-func (s *syncWorkerPool) submitRequests(requests []*messages.BlockRequestMessage) (resultCh chan *syncTaskResult) {
-	resultCh = make(chan *syncTaskResult, maxRequestsAllowed+1)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for task := range failedTasks {
+			if len(workerPool) > 0 {
+				wg.Add(1)
+				go func(t *syncTask) {
+					defer wg.Done()
+					executeTask(t, workerPool, failedTasks, results)
+				}(task)
+			} else {
+				results <- &syncTaskResult{
+					completed: false,
+					request:   task.request,
+					response:  nil,
+				}
+			}
+		}
```

Comment:
I feel like the retry logic can be simplified.  Could you not just make `executeTask` return an error? and you can append to a `tasks` channel which will send a new request.  I feel like reading from the `workerPool` channel should be done before `executeTask`.  Also is the intended behaviour not to retry any failed tasks any of the peers/workers used already?